### PR TITLE
Android: Support for AndroidTV Oreo Homescreen channels

### PR DESCRIPTION
**Pull request by zackhow**
Creates one homescreen channel per game platform. Clicking on the channel logo will launch the main activity. Clicking on a game will launch the emulator and go straight to game. 
screen
screen2
screen3

### CODE REVIEW

**Comment by degasus** [comment]
@gwicks @mahdihijazi billygoat is active on #dolphin-dev, you maybe want to get in contact with him there.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/activities/EmulationActivity.java [hunk 1]
```diff
@@ -158,15 +158,21 @@ public static void launch(FragmentActivity activity, GameFile gameFile, int posi
 		launcher.putExtra(EXTRA_PLATFORM, gameFile.getPlatform());
 		launcher.putExtra(EXTRA_SCREEN_PATH, gameFile.getScreenshotPath());
 		launcher.putExtra(EXTRA_GRID_POSITION, position);
+		Bundle options = new Bundle();
 
-		ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(
+		// Will be null if launched from homescreen
+		if(sharedView != null)
```
Comment (lioncash) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/activities/EmulationActivity.java]: There should be a space after the `if`

**Comment by zackhow** [comment]
Cleaned it up

**Comment by degasus** [comment]
@gwicks @mahdihijazi Do you want to review this PR? Else I'm going to YOLO-merge it :/

**Comment by mahdihijazi** [comment]
@degasus give me till tomorrow evening max and I will go through all Android PRs

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+	}
+
+	/**
+	 * Leanback lanucher requires a uri for poster art, so we take the banner vector,
+	 * make a bitmap, save that bitmap, then return the file provider uri.
+	 */
+	public static Uri buildBanner(GameFile game, Context context)
+	{
+		Uri contentUri = null;
+		int[] vector = game.getBanner();
```
Comment (mahdihijazi) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java]: Check first if the game has a screenshot. GameCube banners are really ugly, I would rather to use the screenshot if it is available, also it will make it more consistent of how Dolphin works.
```

File screenshotFile = new File(URI.create(game.getScreenshotPath()));
		if (screenshotFile.exists()) {
			contentUri = getUriForFile(context, getFilePrivider(context), screenshotFile);
			context.grantUriPermission(LEANBACK_PACKAGE, contentUri, FLAG_GRANT_READ_URI_PERMISSION);
		} else {
		}
```

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+		return true;
+	}
+
+	private long getChannelId(JobParameters jobParameters)
+	{
+		PersistableBundle extras = jobParameters.getExtras();
+		return extras.getLong(TvContractCompat.EXTRA_CHANNEL_ID, -1L);
+	}
+
+	private class SyncProgramsTask extends AsyncTask<Long, Void, Boolean>
```
Comment (mahdihijazi) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java]: static

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	{
+		private Context context;
+		private List<GameFile> updatePrograms;
+
+		private SyncProgramsTask(Context context)
+		{
+			this.context = context.getApplicationContext();
+			updatePrograms = new ArrayList<>();
+			GameFileCacheService.startLoad(context);
```
Comment (mahdihijazi) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java]: Why is this needed? and if it is needed, it doesn't seem we are syncing them together, meaning we are not waiting until the GameFileCacheService to finish before we start this one.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+		return (int) (CHANNEL_JOB_ID_OFFSET + channelId);
+	}
+
+	/**
+	 * Generates all subscriptions for homescreen channels.
+	 */
+	public static List<HomeScreenChannel> createUniversalSubscriptions()
+	{
+		List<HomeScreenChannel> subs = new ArrayList<>();
+		subs.addAll(createPlatformSubscriptions());
```
Comment (mahdihijazi) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java]: Can me simplified with a constructor call passing the list to copy from.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java (hunk 3 above)
Comment (BhaaLseN) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java]: A static class deriving from a different class? Is that a thing in Java, or did you simply miss the `extends AsyncTask<...>` at the end?

Comment (mahdihijazi) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java]: @BhaaLseN yes, that is completely valid and encouraged in Java whenever possible to have a static inner class so you don't have access to the outer class. Two positive things to gain, you won't get access to the outer class reference so you can avoid any possible memory leak from that access. The second thing due to the limited access to the outer class you get clean classes (Inner/Outer) that can be separated later if needed into their own java files.

Comment (BhaaLseN) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java]: Oh, so it's not like a `static class` (cannot create an instance) but more like a visibility/scope feature (reusing the `static` keyword)? It feels strange to create an instance of a class marked as `static`, but it appears that Java has a different meaning for them when they're nested. 👍

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+		return null;
+	}
+
+	/**
+	 * Updates all Leanback homescreen channels
+	 */
+	public static void updateAllChannels(Context context)
+	{
+		List<Channel> channels = getAllChannels(context);
+		for (Channel channel : channels)
```
[the remark this answers is not in the document]
Comment (zackhow) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java]: Fixed.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/ui/main/TvMainActivity.java [hunk 7]
```diff
@@ -53,6 +55,10 @@ protected void onCreate(Bundle savedInstanceState)
 		// Stuff in this block only happens when this activity is newly created (i.e. not a rotation)
 		if (savedInstanceState == null)
 			StartupHandler.HandleInit(this);
+
+		// Setup and/or sync channels
+		if (Build.VERSION.SDK_INT >= 26)
```
[the remark this answers is not in the document]
Comment (zackhow) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/ui/main/TvMainActivity.java]: Fixed both

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+			deletePrograms(channelId);
+			createPrograms(channelId);
+		}
+
+		private void createPrograms(long channelId)
+		{
+			for (GameFile game : updatePrograms)
+			{
+				PreviewProgram previewProgram = buildProgram(channelId, game);
+				if (previewProgram == null)
```
[the remark this answers is not in the document]
Comment (zackhow) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java]: Fixed

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java (hunk 2 above)
Comment (zackhow) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java]: I didn't realize that was the current functionality. I've never actually taken a screenshot in a game before. :) Added.

Also noticed I was creating the banner each time. Added a check to see if it's been done before.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java (hunk 3 above)
Comment (zackhow) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java]: Fixed. I also went ahead and made the Context that is passed a weak reference.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java (hunk 4 above)
Comment (zackhow) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java]: You're right, this isn't needed. Updating the programs only occurs if they open the app or when they add a new game directory. At both points the games cache should already be initialized. Fixed.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java (hunk 5 above)
Comment (zackhow) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java]: So initially I thought about adding a 'Favorites' channel that is platform agnostic, which is why it is how it is, but that would require adding that feature to the main screen then also adding a way persist this group. I figured it should be a separate PR since if its added to TV UI it should be added to the phone UI too and this PR is really just for the Oreo channels. 

I've condensed it for this PR. Can always go back and change it if a 'Favorites' row ever gets added.

**Comment by zackhow** [comment]
Didn't realize Github hid all comments after an update :(.  Either way, fixed all those up. For the 'createUniversalSubscriptions' method, initially I was going to add a Favorites channel for those people who have a ton of games, which would require adding the row to the main activity, then setting up a way to persist the group. But if it was being added to the TV UI, it should be added to the phone UI as well. I figured that should be a separate PR since this is PR for the TV Oreo channels only. So if that happens, it can be changed back.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+	 * Leanback lanucher requires a uri for poster art, so we take the banner vector,
+	 * make a bitmap, save that bitmap, then return the file provider uri.
+	 */
+	public static Uri buildBanner(GameFile game, Context context)
+	{
+		Uri contentUri = null;
+
+		try
+		{
+			//Substring needed to strip "file:" from the path beginning
```
Comment (mahdihijazi) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java]: Are you sure you need to take out the "file:" part, I've tested while using the full path and it worked just fine.

Comment (zackhow) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/TvUtil.java]: Yeah its needed, for some reason the tvlauncher package won't find the file if the URI has the scheme prepended.

**Comment by zackhow** [comment]
Makes sense on the weakreference. I've changed it back.
